proof/refine/ARM/SchedContext_R.thy

```diff
@@ -8,8 +8,521 @@ theory SchedContext_R
+  and ksCurThread[wp]: "\<lambda>s. P (ksCurThread s)"
+  and valid_release_queue[wp]: valid_release_queue
+  and valid_release_queue'[wp]: valid_release_queue'
+  and ct_not_inQ[wp]: ct_not_inQ
+  and valid_pde_mappings'[wp]: valid_pde_mappings'
+  and valid_queues[wp]: valid_queues
+  and ksQ[wp]: "\<lambda>s. P (ksReadyQueues s p)"
+  (wp: crunch_wps simp: crunch_simps tcb_cte_cases_def)
+
+global_interpretation schedContextUpdateConsumed: typ_at_all_props' "schedContextUpdateConsumed scPtr"
+  by typ_at_props'
+
+lemma schedContextUpdateConsumed_if_unsafe_then_cap'[wp]:
+  "schedContextUpdateConsumed scPtr \<lbrace>if_unsafe_then_cap'\<rbrace>"
+  apply (clarsimp simp: schedContextUpdateConsumed_def)
+  apply (wpsimp wp: threadSet_ifunsafe' threadGet_wp)
+  done
+
+lemma schedContextUpdateConsumed_invs'[wp]:
+  "schedContextUpdateConsumed scPtr \<lbrace>invs'\<rbrace>"
+  apply (simp add: invs'_def valid_state'_def valid_pspace'_def setSchedContext_def)
+  apply (wpsimp wp: valid_irq_node_lift valid_irq_handlers_lift'' irqs_masked_lift cur_tcb_lift
+                    untyped_ranges_zero_lift
+              simp: cteCaps_of_def o_def)
+  done
+
+lemma schedContextCancelYieldTo_valid_objs'[wp]:
+  "schedContextCancelYieldTo tptr \<lbrace>valid_objs'\<rbrace>"
+  apply (clarsimp simp: schedContextCancelYieldTo_def)
+  apply (rule hoare_seq_ext[OF _ threadGet_sp])
+  apply (rule hoare_when_cases; (solves \<open>wpsimp\<close>)?)
+  apply (rule hoare_seq_ext[OF _ get_sc_sp'])
+  apply (rule_tac B="\<lambda>_. valid_objs'" in hoare_seq_ext[rotated])
+   apply (wpsimp wp: set_sc_valid_objs')
+   apply (fastforce dest!: obj_at_valid_objs'
+                     simp: projectKOs valid_obj'_def valid_sched_context'_def
+                           valid_sched_context_size'_def scBits_simps objBits_simps')
+  apply (wpsimp wp: threadSet_valid_objs')
+  apply (clarsimp simp: valid_tcb'_def tcb_cte_cases_def)
+  done
+
+lemma schedContextCancelYieldTo_valid_mdb'[wp]:
+  "schedContextCancelYieldTo tptr \<lbrace>valid_mdb'\<rbrace>"
+  apply (clarsimp simp: schedContextCancelYieldTo_def threadSet_def )
+  apply (wpsimp wp: getObject_tcb_wp hoare_drop_imps hoare_vcg_ex_lift threadGet_wp)
+  apply (fastforce simp: obj_at'_def projectKOs tcb_cte_cases_def)
+  done
+
+lemma schedContextCancelYieldTo_sch_act_wf[wp]:
+  "schedContextCancelYieldTo tptr \<lbrace>\<lambda>s. sch_act_wf (ksSchedulerAction s) s\<rbrace>"
+  apply (clarsimp simp: schedContextCancelYieldTo_def)
+  apply (wpsimp wp: threadSet_sch_act threadGet_wp)
+  apply (fastforce simp: obj_at'_def projectKOs)
+  done
+
+lemma schedContextCancelYieldTo_if_live_then_nonz_cap'[wp]:
+  "\<lbrace>\<lambda>s. if_live_then_nonz_cap' s\<rbrace>
+   schedContextCancelYieldTo tptr
+   \<lbrace>\<lambda>_. if_live_then_nonz_cap'\<rbrace>"
+  apply (clarsimp simp: schedContextCancelYieldTo_def)
+  apply (rule hoare_seq_ext[OF _ threadGet_sp])
+  apply (rule hoare_when_cases; (solves \<open>wpsimp\<close>)?)
```

Comment:
Is this okay to have in the final proof? I guess it means apply `wpsimp` to every produced subgoal that can be solved by `wpsimp` otherwise leave it alone.

Reply:
Yes, that's what it means and should be fine.